File: java/leetcode/src/main/java/com/leetcode/tree/Path_Sum_112.java
package com.leetcode.tree;

public class Path_Sum_112 {
    boolean found = false;
    public boolean hasPathSum(TreeNode root, int targetSum) {
        dfs(root, targetSum);
        return found;
    }

    void dfs(TreeNode node, int left) {
        if (node == null) return;
        if (isLeaf(node)) {
            if (node.val == left) found = true;
        } else {
            dfs(node.left, left - node.val);
            if (!found) {
                dfs(node.right, left - node.val);
            }
        }
    }

    boolean isLeaf(TreeNode node) {
        return node.left == null && node.right == null;
    }
}

File: js/algrithm/huahua130/tree/Path_Sum_112.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {
    let flag = false
    const dfs = function(node, sum) {
        if (!node) return
        if (!node.left && !node.right) {
            if (sum + node.val === targetSum) flag = true
        }
        dfs(node.left, sum + node.val)
        if (!flag) {
            dfs(node.right, sum + node.val)
        }
    }

    dfs(root, 0)
    return flag
};